src/main/java/uk/gov/hmcts/reform/sendletter/tasks/SerialTaskRunner.java

```diff
@@ -0,0 +1,85 @@
+package uk.gov.hmcts.reform.sendletter.tasks;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import javax.sql.DataSource;
+
+/**
+ * Runs tasks in serial using Database locks
+ * to ensure the same task is not run concurrently.
+ */
+public final class SerialTaskRunner {
+
+    private static final Logger log = LoggerFactory.getLogger(SerialTaskRunner.class);
+    private final DataSource source;
+
+    public static SerialTaskRunner get(DataSource source) {
+        return new SerialTaskRunner(source);
+    }
+
+    private SerialTaskRunner(DataSource source) {
+        this.source = source;
+    }
+
+    /**
+     * Run the task with specified ID if no other task with matching
+     * id is already running.
+     * No error is thrown if such a task is already running, the
+     * supplied Runnable is simply not executed.
+     */
+    public void tryRun(long id, Runnable runnable) throws SQLException {
```

Comment:
How would this be invoked? Will tasks/jobs have fixed ids assigned to them?

Reply:
Yes I was thinking having something like a 'TaskLock' enum, where each of our tasks have a value.

Another approach would be to just have a single scheduled task that wraps all of our subtasks, then none of our tasks would run concurrently (but we might not want all our tasks running on the same schedule).